src/main/scala/latis/service/hapi/DataService.scala

```diff
@@ -25,33 +26,44 @@ class DataService[F[_]: Concurrent](
   val service: HttpRoutes[F] =
     HttpRoutes.of[F] {
       case GET -> Root / "data"
-          :? IdMatcher(_id)
+          :? DatasetMatcher(_dataset)
+          +& IdMatcher(_id)
+          +& StartTimeMatcher(_startTime)
+          +& StopTimeMatcher(_stopTime)
           +& MinTimeMatcher(_minTime)
           +& MaxTimeMatcher(_maxTime)
           +& ParamMatcher(_params)
           +& IncludeMatcher(_inc)
           +& FormatMatcher(_fmt) =>
-        val req: Either[Status, DataRequest] =
+        val req: Either[Status, DataRequest] = {
+          // For backwards compatibility, allow both
+          // "dataset" or "id",
+          // "start" or "time.min",
+          // "stop" or "time.max"
           for {
-            id      <- _id.asRight
-            minTime <- _minTime.leftMap(_ => Status.`1402`).toEither
-            maxTime <- _maxTime.leftMap(_ => Status.`1403`).toEither
-            _       <- Either.cond(minTime.isBefore(maxTime), (), Status.`1404`)
-            params  <- Either.cond(
+            dataset   <- _dataset.getOrElse(_id.getOrElse("")).asRight
```

Comment:
Some thoughts on this line:

Here we're trying to state that we want the value of `dataset` if it is defined, the value of `id` if it's not, or to fail with a 1400 if `id` isn't defined either (because at least one of `dataset` and `id` must be defined).

You can think of this as having two `Option`s and needing to combine them together to get a single `Option`. Luckily, we know about something that can help us combine two of the same thing together: semigroups! Specifically, the [`SemigroupK` typeclass](https://typelevel.org/cats/typeclasses/semigroupk.html).

The distinction between `Semigroup` and `SemigroupK` for `Option` is that `Semigroup` combines the things inside the `Option` while `SemigroupK` combines the `Option`s themselves. The default way to combine two `Option`s is to take the first `Some`, which is exactly what we want here:

```
@ Option("a") <+> None 
res3: Option[String] = Some("a")

@ Option("a") <+> Option("b") 
res4: Option[String] = Some("a")

@ None <+> Option("b") 
res5: Option[String] = Some("b")
```

Compare that to the `Semigroup` instance:

```
@ Option("a") |+| None 
res1: Option[String] = Some("a")

@ Option("a") |+| Option("b") 
res2: Option[String] = Some("ab")

// Can't use None directly because we need to know the type of what the Option contains.
// It worked with SemigroupK because we don't care about the contents.
@ Option.empty[String] |+| Option("b") 
res4: Option[String] = Some("b")
```

The `<+>` operator is named `combineK` and is the same thing as the `orElse` method for `Option`s.

This gets us to here:

```scala
(_dataset <+> _id).getOrElse("").asRight
```

However, we want to fail if `(_dataset <+> _id)` is `None`, because that implies that neither were set in the query. So we want to lift our `Option` into `Either` and provide the appropriate value to put in `Left` if our `Option` is `None`. That's what the `Either.fromOption` method is for:

```scala
Either.fromOption(_dataset <+> _id, Status.`1400`)
```

So this says "give me the value of `_dataset` if it's defined, `_id` if it's not defined, and fail with a 1400 status if neither are defined."

Reply:
Thank you. 🙏